Feat/setup talk2comp

ansh-info opened the pull request:
# For authors

https://github.com/user-attachments/assets/87c1556a-4062-467b-88c2-aa0bed7ac628

## Description

Hello, my name is Ansh Kumar. I am a student at SRH University and am currently working on a case study in collaboration with the Team VPE on Talk2Competitors. I was part of the winning team at the Biodatathon.

#### Proposed Agentic architecture of Talk2Competitors
<img width="1346" alt="image" src="https://github.com/user-attachments/assets/3993c326-c5dc-4cd0-96eb-2971ac06234a" />


This PR implements the initial version of our hierarchical agent system focused on academic paper search and management. The implementation includes a main supervisor agent and a subagent - Semantic Scholar (S2) integration using LangGraph. The Main Supervisor Agent and Semantic Scholar (S2) Agent is built using ReAct pattern([create_react_agent](https://langchain-ai.github.io/langgraph/how-tos/create-react-agent/#code))

This implementation is inspired by LangGraph's [Hierarchical Agent Teams](https://langchain-ai.github.io/langgraph/tutorials/multi_agent/hierarchical_agent_teams/)


The s2 agent have the following tools:

- search_tool: Paper search based on key-words
- single_paper_recommendations: Recommending articles against a given article
- multi_paper_recommendations: Recommending articles against a given group of articles
- display_results: Display articles


Logging and LangSmith are also integrated for tracing purposes.
For UI I followed the streamlit application code that we have over at AIAgents4Pharma for other agents.

## Subsequent PR's Improvements
1. Using Hydra for configs(Currently I am using config.py which located in [config/](https://github.com/ansh-info/AIAgents4Pharma/tree/feat/setup-talk2comp/aiagents4pharma/talk2competitors/config) for my system prompt. I have already commented on issue #61)
2. Implementation of Talk2Competitors agents (Zotero, PDF, arXiv/PMC)
3. Adding Additional arguments to the tools in order to make more specific queries to the application
4. Add more query parameter in the search tools for agent s2 for example: abstract, citation count 
5. ollama/Base_Chat_Model
6. Use Tool_Exception for API/error handling of tools
7. Utility function to create LangGraph agent using a given set of tools
8. A Visualisation agent to render images/dataframes

## Fixes # (issue) Mention the issue number. -  NA


## Type of change
Please delete options that are not relevant.
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## How Has This Been Tested?
Please describe the tests you conducted to verify your changes. These may involve creating new test scripts or updating existing ones.
- [x] Added new test(s) in the ```tests``` folder
- [ ] Added new function(s) to an existing test(s) (e.g.: ```tests/testX.py```)
- [ ] No new tests added (Please explain the rationale in this case)

## Checklist
- [x] My code follows the style guidelines mentioned in the Code/DevOps guides
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (e.g. MkDocs)
- [x] My changes generate no new warnings
- [x] I have added or updated tests (in the ```tests``` folder) that prove my fix is effective or that my feature works
- [x] New and existing tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

# For reviewers

## Checklist pre-approval
- [ ] Is there enough documentation?  
- [ ] If a new feature has been added, or a bug fixed, has a test been added to confirm good behavior?
- [ ] Does the test(s) successfully test edge/corner cases?
- [ ] Does the PR pass the tests? (if the repository has continuous integration)

## Checklist post-approval
- [ ] Does this PR merge ```develop``` into ```main```? If so, please make sure to add a prefix (feat/fix/chore) and/or a suffix BREAKING CHANGE (if it's a major release) to your commit message.
- [ ] Does this PR close an issue? If so, please make sure to descriptively close this issue when the PR is merged.

## Checklist post-merge
- [ ] When you approve of the PR, merge and close it (Read this [article](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/about-merge-methods-on-github) to know about different merge methods on GitHub)
- [ ] Did this PR merge ```develop``` into ```main``` and is it suppose to run an automated release workflow (if applicable)? If so, please make sure to check under the "Actions" tab to see if the workflow has been initiated, and return later to verify that it has completed successfully.

dmccloskey commented on app/frontend/streamlit_app_talk2competitors2.py:
What is the reason for the second app?

ansh-info replied:
I apologize, I accidentally pushed the file to the repository.

awmulyadi commented:
Good job, Ansh, for your first PR! :1st_place_medal:   I have reviewed your code and run the app briefly, which leaves me with several concerns/questions.

Please find my comments below:
- `aiagents4pharma/talk2competitors/config/config.py`: I assumed that the current version of the app only implements a semantic scholar agent to search for and recommend papers. However, the `MAIN_AGENT_PROMPT` in the config.py contains prompts with other agents and more advanced capabilities beyond searching and recommendation. I would recommend modifying the prompt to reflect the current implemented features of talk2comp.
    ![Screenshot From 2025-01-23 10-47-06](https://github.com/user-attachments/assets/2e936a08-7045-4748-add8-0b11c3d577dc)

- I have tried the use case where the prompt specifically asks for recent papers (e.g., 2024 onward). However, there are still some papers that are way older than the instructions. Would it be possible for the current app to cope with this specific use case? Or rather, it is tied to the LLM model to be run with?

    ![Screenshot From 2025-01-23 10-17-25](https://github.com/user-attachments/assets/4ca0b101-3b6a-4860-be01-47458bc977a5)

    Here is the year of papers when I did a manual search in Semantic Scholar:
    1. Assessing Scientific Research Papers with Knowledge Graphs (2022)
    2. Advances on the theory of M&S (2013)
    3. A Survey on Self-Supervised Graph Foundation Models: Knowledge-Based Perspective (2024)
    4. Towards Graph Foundation Models: A Study on the Generalization of Positional and Structural Encodings (2024)
    5. Lecture-style Tutorial: Towards Graph Foundation Models (2024)

- It seems that there is no standard format to display the results. I have captured the result where there is no hyperlink appearing for each paper in the results, as shown below.

    Result with hyperlinks:
    ![Screenshot From 2025-01-23 10-33-52](https://github.com/user-attachments/assets/68c262e6-afa3-40e0-83a8-36fa9e693a3e)

    Result without hyperlinks:
    ![Screenshot From 2025-01-23 10-34-48](https://github.com/user-attachments/assets/2e63517d-970a-4183-949e-0059df2e2aac)

    Feel free to discuss further with me if my comments are not clear enough. Also, please ignore my comments if you think they will not fit in the current PR or are supposed to be resolved in the next PRs. :+1:

ansh-info commented on aiagents4pharma/talk2competitors/config/config.py:
```diff
@@ -0,0 +1,142 @@
+    SEMANTIC_SCHOLAR_API_KEY = "YOUR_API_KEY"  # Get this from Semantic Scholar
+
+    # Updated System Prompts
+    MAIN_AGENT_PROMPT = """You are a supervisory AI agent that routes user queries to specialized tools.
```
Hello,

Thank you so much for your valuable feedback! I truly appreciate you taking the time to run my application on your machine and sharing your insights—it means a lot to me.

I’ve carefully reviewed all the points you mentioned and addressed them as follows:

- I have updated the system prompt to accurately reflect the currently implemented features of Talk2Comp.
- I’ve added the capability to specifically query for recent papers (e.g., 2024 onward), as per your request.
- Additionally, I’ve introduced a standardized display format, ensuring that results include hyperlinks for easier navigation.

To assist with your review, I’ve also included a demo video using a query similar to the one you provided. Could you please review the changes when you get a chance?


https://github.com/user-attachments/assets/9d56db46-5ac9-42e9-a4fe-5275146c2fea

Additionally, I have included the year and abstract in the search results for all three tools.

Thank you once again for your time and support!

awmulyadi replied:
Looks great after the most recent updates of the app! The current behavior of the app in recommending the papers is as expected. In addition, it was a good feature to include the abstract, published year, as well as citation counts of the paper. Well done! :+1:

ansh-info commented on aiagents4pharma/talk2competitors/tests/test_langgraph.py:
```diff
@@ -0,0 +1,92 @@
+
+    assistant_msg = response["messages"][-1].content
+    # Check if the assistant message is a string
+    assert 'Fashion-MNIST' in assistant_msg
```
Thank you for taking the time to review my code and for the kind welcome! Your feedback is invaluable in improving the code quality.

Regarding the LLM dependency concern, I want to clarify that the output is actually deterministic in our implementation. The LLM's role is limited to routing and doesn't influence the paper search results. The papers are directly fetched from the Semantic Scholar API and stored in a structured dictionary format in our shared state. 

In fact, when searching for "machine learning," you get consistent results from Semantic Scholar API - the "Fashion-MNIST" paper consistently appears as a top result (I tested this manually as well). However, I completely agree that we can make our tests even more deterministic.

I've updated the tests to:
1. Use fixed mock responses with known ML-related content
2. Verify exact paper structure and content
3. Test state management independently
4. Include assertions for data format and consistency

Would you like me to make any additional improvements to enhance test determinism?

Please also have a look at this screenshot 
<img width="1445" alt="image" src="https://github.com/user-attachments/assets/71d409dc-6f35-4804-8cca-1d4fd0bc8713" />

ansh-info replied:
Regarding the documentation, I am already reviewing the repository and currently working on the Jupyter notebook section as well as the MKDocs part. I will update you as soon as I have completed these tasks.

dmccloskey replied:
The tests look much better. Nice work 👍. Ping me for re-review once you have added the jupyter notebooks and mkdocs.

ansh-info commented on aiagents4pharma/talk2competitors/tests/test_langgraph.py:
```diff
@@ -0,0 +1,92 @@
+    # Check if the assistant message is a string
+    assert 'Fashion-MNIST' in assistant_msg
+    ####################################################
+    prompt = "Recommend articles using the first paper of the previous search"
```
Thank you for this important feedback about test dependencies. You're absolutely right - I've refactored the tests to follow better testing practices.

The updated test suite now includes:
1. Independent unit tests for each component:
   - Main agent routing
   - Individual S2 tools (search, recommendations, display)
   - State management
2. Isolated test cases with their own setup
3. A separate integration test for the complete workflow
4. No dependencies between test cases

Each test now has its own mock data and assertions, making them more reliable and easier to maintain. Thank you for helping me improve the test architecture.

dmccloskey replied:
The testing structure looks much better 👍

ansh-info commented on aiagents4pharma/talk2competitors/tests/test_langgraph.py:
```diff
@@ -0,0 +1,92 @@
+        config=config,
+    )
+
+    assistant_msg = response["messages"][-1].content
```
Thank you for highlighting the need for clearer test functionality documentation. I've restructured the tests to make their purposes explicit and map directly to functionality:

1. TestMainAgent class:
   - Tests supervisor routing logic
   - Verifies correct agent selection
   - Checks state management

2. TestS2Tools class:
   - Tests each tool's core functionality
   - Verifies data transformation
   - Validates response formats

3. Integration test:
   - Verifies end-to-end workflow
   - Tests component interaction

Each test now has detailed docstrings explaining what's being tested and why. This one-to-one correspondence between functionality and tests makes it clearer what each test is verifying.

Would you like me to add any additional documentation to make the test purposes even clearer?

ansh-info replied:
Can you also please check the updated demo below, and provide your feedback please.
I’ve added the capability to specifically query for recent papers (e.g., 2024 onward), as per the request.

dmccloskey replied:
What functionality and what specific cases are being tested is much much clearer 😊. @awmulyadi can comment on the recent papers part.

dmccloskey reviewed (commented):
Much improved 👏. @ansh-info and @gurdeep330 will there also be updated MKDocs and Jupypter notebook examples in this PR?

ansh-info commented:
Hello @dmccloskey about the MKDocs and Jupyter notebook example : This will actually be covered in the next PR. I've already opened an issue #70 to track it and plan to work on it after my semester exams, which will conclude by early February. I expect to complete this between the 7th and 10th of February.

ansh-info commented:
> Hello @dmccloskey about the MKDocs and Jupyter notebook example : This will actually be covered in the next PR. I've already opened an issue #70 to track it and plan to work on it after my semester exams, which will conclude by early February. I expect to complete this between the 7th and 10th of February.

@dmccloskey I wanted to inform you that I have completed this task earlier than expected. Could you kindly review it at your convenience? Your feedback would be greatly appreciated.

dmccloskey commented on docs/notebooks/talk2competitors/tutorial.ipynb:
Very nice 🤩. The tutorial does a nice job of walking through the functionality.

Can you provide a complete example for loading the API key? e.g., `load_dotenv()`: can you provide a dummy path and how one would actually load the variables into their script?

ansh-info replied:
@dmccloskey  I’ve updated the tutorial to include a complete example for loading the API key. It now demonstrates how to use load_dotenv() with a dummy path to the .env file and provides step-by-step instructions for loading variables into the script.

When you have a moment, could you please take a look and let me know if there’s anything else you'd like me to adjust? 😊

Thanks again for your input—it’s much appreciated!

dmccloskey commented on docs/talk2competitors/intro.md:
```diff
@@ -0,0 +1,29 @@
+
+Academic literature search can be overwhelming, with millions of papers published yearly across diverse fields. **Talk2Competitors** transforms this experience by offering an intelligent, conversation-driven approach to academic research. Our system acts as a knowledgeable research assistant that understands context and provides targeted recommendations.
+
+At its core, Talk2Competitors uses a sophisticated hierarchical agent architecture. The main supervisor agent intelligently routes queries to specialized sub-agents, ensuring that each request is handled by the most appropriate tool. Currently, our Semantic Scholar integration provides deep access to academic literature, with plans to expand to additional databases and features.
```
```suggestion
At its core, Talk2Competitors uses a sophisticated hierarchical agent architecture. The main supervisor agent intelligently routes queries to specialized sub-agents, ensuring that each request is handled by the most appropriate tool. Currently, our Semantic Scholar integration provides access to academic literature, which we plan to expand in the future.
```

dmccloskey commented on docs/talk2competitors/intro.md:
```diff
@@ -0,0 +1,29 @@
+### Key Capabilities
+
+The system excels at understanding natural language queries and translating them into precise academic searches. When you find interesting papers, Talk2Competitors can analyze them to discover related research, either based on a single paper or by finding connections between multiple papers. This helps researchers uncover hidden connections in literature and stay current with relevant publication ...
+
```
Please add as an additional line for this section.

`We are in the process integrating [Zotero](https://www.zotero.org/) to enable researchers to better organize their literature searchers, get automated recommendations based on categories of papers, and cite saved papers when asking questions about one or more saved papers or generating text based on a starting prompt.`

ansh-info replied:
Added a note about integrating Zotero to enhance literature management and citation support.

dmccloskey reviewed (changes requested):
Updates to the documentation and notebook look good 👏.

Just a few minor comments from my side to improve the documentation and notebook.

dmccloskey reviewed (approved):
Looks good 👍. Now you just need to convince @awmulyadi 😉.

awmulyadi reviewed (approved):
Looks good to me for the initial setup and basic features of talk2competitors. Great work! :1st_place_medal:

ansh-info commented:
> Hi Ansh! 👋 Thanks for your efforts implementing the Talk2Competitors agent! 🥳 Here are my comments and suggestions to current state of the functionality:
> 
> * If a user would like display multiple articles, it would be convenient to view the results in  a table with **author names**, and **h-index**, as seen ,e.g., here (https://virtualpatientengine.github.io/literatureSurveyModels/MAPK/)
> * The **sorting** of the table rows by date/name/citations etc. would be extremely convenient.
> * This search enquiry ("show me the latest articles on physics-based graph neural networks") **did not return any results or an error message** (see below). However, S2 search works fine: https://www.semanticscholar.org/search?q=physics-based%20graph%20neural%20networks&sort=relevance
> * Also search **using multiple keywords** does not return any result:
>   ![image](https://private-user-images.githubusercontent.com/43205236/406910102-a9603598-4d69-4326-bb79-a1147ad03cbe.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzc5ODUyMjQsIm5iZiI6MTczNzk4NDkyNCwicGF0aCI6Ii80MzIwNTIzNi80MDY5MTAxMDItYTk2MDM1OTgtNGQ2OS00MzI2LWJiNzktYTExNDdhZDAzY2JlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTI3VDEzMzUyNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJiYTBkNjdlODA5MmE2MjA3ZmE3MDhmMzYxMzc4OTY3MDdjNDEyMzcxZmIwYWM0MWYwNGViZTMzYzgwNzk1ZjAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.CxeT2RKodAQtz8S6iVlK9tYjPbvomGqj9RhQjlxGjWk) Is this expected behavior?
> 
> Feel free to discuss further, if you have any questions or comments regarding these points. Also, please feel free to address these issues in the next PRs. Let me know your thoughts!

Here’s a polished response to the review:

---

Hi @lilijap  👋

Thank you so much for taking the time to review the PR and run the implementation on your end. I really appreciate your feedback and detailed observations! 🙏

### Addressing Your Points:

1. **Displaying Results in a Table with Sorting Options**:
   - I’ve already started working on this feature. The plan is to fetch the papers directly from the shared state, convert them into a Streamlit dataframe, and display them in a table format. 
   - I’ll also include sorting functionality for table rows (by date, author name, citations, etc.) to make it more user-friendly. 
   - However, as I’m currently preparing for my semester exams, would it be okay to implement this in the next PR? I’ll try to work on it alongside my studies as much as possible. I’ve also opened an [issue](https://github.com/VirtualPatientEngine/AIAgents4Pharma/issues/76)  to track this enhancement.

2. **Search Queries Not Returning Results**:
   - I believe the LLM is getting confused with the term **"show"** in the query. The `create_react_agent` assumes we have a fourth tool (display results) and attempts to fetch data directly from the shared state instead of executing a fresh search. 
   - If you use the term **"find"** instead of **"show"**, the search queries work as expected for both cases (single and multiple keywords). I’ve attached screenshots demonstrating this behavior.
   - This isn’t a bug per se but rather the system trying to identify the most appropriate tool for the query.
   
3. **Future Improvements**:
   - In the next PR, we plan to transition to displaying results directly in a table, fetching raw data from the shared state. This will allow us to completely remove `display_results.py`, as it will no longer be required.
   - To address the confusion between "show" and "find," we can either update the system prompt to clarify this behavior or introduce an error message in the frontend. For example, we could display a message like:  
     _"We don’t have the related query paper in the state. Did you mean to find it instead?"_

#### "Show/Find"
<img width="1668" alt="image" src="https://github.com/user-attachments/assets/a09f72d8-2e3d-463c-aff1-8f8455a76566" />

## Multiple Keywords
<img width="1649" alt="image" src="https://github.com/user-attachments/assets/32919dd2-6241-47df-8d40-064527ee7f9a" />


Thanks again for your valuable input! Feel free to share any additional feedback or questions—I’m happy to discuss further. 😊
